tools/benchmark/src/utilities.ts

```diff
@@ -0,0 +1,185 @@
+
+	/**
+	 * Arithmetic mean.
+	 */
+	readonly arithmeticMean: number;
+
+	/**
+	 * Sample values.
+	 */
+	readonly samples: readonly number[];
+
+	/**
+	 * Variance.
+	 */
+	readonly variance: number;
+}
+
+/**
+ * Contains the samples of all memory-related measurements we track for a given benchmark (a test which was
+ * potentially iterated several times). Each property is an array and all should be the same length, which
+ * is the number of iterations done during the benchmark.
+ * @public
+ */
+export interface MemoryTestData {
+	/**
+	 * Memory usage in bytes.
+	 */
+	memoryUsage: NodeJS.MemoryUsage[];
+
+	/**
+	 * Heap info.
+	 */
+	heap: v8.HeapInfo[];
+
+	/**
+	 * Heap space info.
+	 */
+	heapSpace: v8.HeapSpaceInfo[][];
+}
+
+/**
+ * Contains the samples of all memory-related measurements before and after a benchmark.
+ * @public
+ */
+export interface MemorySampleData {
+	/**
+	 * Memory usage before the test.
+	 */
+	before: MemoryTestData;
+
+	/**
+	 * Memory usage after the test.
+	 */
+	after: MemoryTestData;
+}
+
+/**
+ * Result of successfully running a benchmark.
+ * @public
+ */
+export interface BenchmarkData {
+	/**
+	 * Iterations per batch.
+	 */
+	readonly iterationsPerBatch: number;
+
+	/**
+	 * Number of batches, each with `iterationsPerBatch` iterations.
+	 */
+	readonly numberOfBatches: number;
+
+	/**
+	 * Stats about runtime, in seconds.
+	 * This is already scaled to be per iteration and not per batch.
+	 */
+	stats: Stats;
+
+	/**
+	 * Time it took to run the benchmark in seconds.
+	 */
+	readonly elapsedSeconds: number;
+
+	/**
+	 * Custom data.
+	 */
+	customData: Record<string, unknown>;
+}
+
+/**
+ * Result of failing to run a benchmark.
+ * @public
+ */
+export interface BenchmarkError {
+	/**
+	 * Error message.
+	 */
+	error: string;
+}
+
+/**
+ * Contains the full results for a benchmark (a test which was potentially iterated several times).
+ * 'samples' contains the raw 'before' and 'after' measurements instead of calculated deltas
+ * for flexibility, at the cost of more memory and bigger output.
+ * @public
+ */
+export interface MemoryBenchmarkStats {
+	/**
+	 * Number of runs.
+	 */
+	runs: number;
+
+	/**
+	 * Contains the samples of all memory-related measurements we track for a given benchmark (a test which was
+	 * potentially iterated several times). Each property is an array and all should be the same length, which
+	 * is the number of iterations done during the benchmark.
+	 */
+	samples: MemorySampleData;
+
+	/**
+	 * Stats about the arithmetic mean of the delta between the 'before' and 'after' measurements.
+	 */
+	stats: Stats;
+
+	/**
+	 * Whether the benchmark was aborted.
+	 */
+	aborted: boolean;
+
+	/**
+	 * Error message.
+	 */
+	error?: Error;
+
+	/**
+	 * Total run time in milliseconds.
+	 */
+	totalRunTimeMs: number;
+}
+
+/**
+ * Result of trying to run a benchmark.
+ * @public
+ */
+export type BenchmarkResult = BenchmarkError | Partial<BenchmarkData>;
```

Comment:
Curious why we made this `Partial<>`? Feels to me like any non-error benchmark should have all of the properties defined in `BenchmarkData`.

Reply:
Seems like to move fields in `MemoryBenchmarkStats` into customData Map of `BenchmarkData`. There are only one common field between them(`Stats`). Moreover, any possible information we want to add in the future, we can just add into `BenchmarkData` and `BenchmarkResult` will still be able to handle in all runners.